src/toplevel-define-type.lisp

```diff
@@ -28,48 +28,50 @@
                 :newtype (type-definition-newtype parsed-deftype)
                 :docstring (type-definition-docstring parsed-deftype))))
 
-        (dolist (ctor (type-definition-constructors parsed-deftype))
-          ;; Add the constructors to the constructor environment
-          (setf env
-                (set-constructor
-                 env
-                 (constructor-entry-name ctor) ctor))
+        (loop :for ctor :in (type-definition-constructors parsed-deftype)
+              :for ctor-type :in (type-definition-constructor-types parsed-deftype)
+              :for ctor-name := (constructor-entry-name ctor)
 
-          ;; Add the constructor as a value to the value environment
-          (setf env
-                (set-value-type
-                 env
-                 (constructor-entry-name ctor)
-                 (constructor-entry-scheme ctor)))
+              ;; Add the constructors to the constructor environment
+              :do
+                 (setf env
+                       (set-constructor
+                        env
+                        (constructor-entry-name ctor) ctor))
 
-          ;; Register the constructor in the name environment
-          (setf env
-                (set-name
-                 env
-                 (constructor-entry-name ctor)
-                 (make-name-entry
-                  :name (constructor-entry-name ctor)
-                  :type :constructor
-                  :docstring nil
-                  :location (or *compile-file-pathname*
-                                *load-truename*))))
+                 ;; Add the constructor as a value to the value environment
+                 (setf env
+                       (set-value-type
+                        env
+                        ctor-name
+                        ctor-type))
 
-          ;; If the constructor takes paramaters then add it to the function environment
-          (if (not (= (constructor-entry-arity ctor) 0))
-              (setf env
-                    (set-function
-                     env
-                     (constructor-entry-name ctor)
-                     (make-function-env-entry
-                      :name (constructor-entry-name ctor)
-                      :arity (constructor-entry-arity ctor))))
+                 ;; Register the constructor in the name environment
+                 (setf env
+                       (set-name
+                        env
+                        (constructor-entry-name ctor)
+                        (make-name-entry
+                         :name (constructor-entry-name ctor)
+                         :type :constructor
+                         :docstring nil
+                         :location (or *compile-file-pathname*
+                                       *load-truename*))))
 
-              ;; TODO: This does not seem right... Is this to ensure
```

Comment:
@gefjon I deleted your TODO comment. Your assumption about redefined constructors is correct.